src/NuGet.Clients/NuGet.CommandLine/MsBuildToolsetEx.cs

```diff
@@ -0,0 +1,174 @@
+
+            if (Version.TryParse(version, out result))
+            {
+                return result;
+            }
+            else
+            {
+                return new Version(0, 0);
+            }
+        }
+
+        public string ToolsPath { get; private set; }
+
+        public static IEnumerable<MsBuildToolsetEx> AsMsToolsetExCollection(IEnumerable<Toolset> toolsets)
+        {
+            return toolsets.Select(t => new MsBuildToolsetEx(t));
+        }
+
+        public int CompareTo(object obj)
+        {
+            if (Object.ReferenceEquals(obj, null))
+            {
+                return 1;
+            }
+
+            var rhs = obj as MsBuildToolsetEx;
+            if (obj == null)
+            {
+                throw new ArgumentException("Comparison object not of correct type");
+            }
+
+            // Compare versions
+            if (this.ParsedToolsVersion.Major > rhs.ParsedToolsVersion.Major)
+            {
+                return 1;
+            }
+            if (this.ParsedToolsVersion.Major < rhs.ParsedToolsVersion.Major)
+            {
+                return -1;
+            }
+
+            // Major versions equal; compare minor versions
+            if (this.ParsedToolsVersion.Minor > rhs.ParsedToolsVersion.Minor)
+            {
+                return 1;
+            }
+            if (this.ParsedToolsVersion.Minor < rhs.ParsedToolsVersion.Minor)
+            {
+                return -1;
+            }
+
+            // Versions equal; compare by install date/time
+            return this._installDateTime > rhs._installDateTime ? 1 : -1;
+        }
+
+        private static DateTime ConvertFILETIMEToDateTime(FILETIME time)
+        {
+            long highBits = time.dwHighDateTime;
+            highBits = highBits << 32;
+            return DateTime.FromFileTimeUtc(highBits | (long)(uint)time.dwLowDateTime);
+        }
+
+        public static string GetMSBuildPathFromVsPath(string vsPath)
+        {
+            if (string.IsNullOrEmpty(vsPath))
+            {
+                return null;
+            }
+
+            string msBuildRoot = Path.Combine(vsPath, "MSBuild");
+            if (!Directory.Exists(msBuildRoot))
+            {
+                return null;
+            }
+
+            // Enumerate all versions of MSBuild present, take the highest
+            string msBuildPath = string.Empty;
+            var highestVersionRoot = Directory.EnumerateDirectories(msBuildRoot).OrderByDescending(dir =>
+            {
+                var dirName = new DirectoryInfo(dir).Name;
+                float dirValue;
+                if (float.TryParse(dirName, out dirValue))
+                {
+                    return dirValue;
+                }
+
+                return 0F;
+            })
+            .FirstOrDefault(dir =>
+            {
+                msBuildPath = Path.Combine(dir, "bin");
+                return File.Exists(Path.Combine(msBuildPath, "msbuild.exe"));
```

Comment:
Consider a situation where there is one MSBuild installation, but msbuild.exe does not exist there.  This method would return the path anyway.  (`highestVersionRoot` isn't observed.)  Is this correct behavior?